Issues: add CRUD on issues

FaiyazMujawar opened the pull request:
Fixes #14 

In this PR:
- Implemented database schema for issues.
- Implemented service layer for issues.
- Implemented controller for issues. 

**Note:** I've had some problems with unit tests. Some tests have not been written, as mentioned in the code with cause. Also, some tests cases are failing, and hence have been commented out. I've marked the failing tests with `BUG` and mentioned the error message with them.

drigio commented on models/Issue.js:
```diff
@@ -0,0 +1,58 @@
+    type: Number,
+  },
+  upvoters: {
+    type: [Number],
```
I guess this is a list of all the `zprn` that have upvoted the issue right?

FaiyazMujawar replied:
Yes, sir.

drigio commented on models/Issue.js:
```diff
@@ -0,0 +1,58 @@
+    type: [Number],
+  },
+  comments: {
+    type: [String],
```
`comments` should also be an `object` just like solution. We need to store who commented, when, etc. Although is information is just for the `db` and is not exposed via the `rest api`.

FaiyazMujawar replied:
Okay, maybe we can make this change in the next PR?

drigio replied:
It'll create a lot of regressions in a later pr. So we should update it in this PR itself in a new commit i guess.

FaiyazMujawar replied:
Okay.

drigio commented on models/Solution.js:
```diff
@@ -0,0 +1,9 @@
+  constructor(userID, solution) {
+    this.postedBy = userID;
+    this.solution = solution;
+    this.postedOn = new Date().toLocaleString();
```
We should have `upvotes`, `isEdited`, `isDeleted`, `isInappropriate` fields as well for the solution schema.

FaiyazMujawar replied:
Are we allowing the users to update the comments/solutions?

drigio replied:
No we should just allow solutions to be updated. If comments are to be updated then that would expose user info to the rest api which we don't want.

FaiyazMujawar replied:
Agree on that. But allowing solutions to be edited wouldn't expose the user info to the API?

drigio replied:
`Solutions` can only be posted by the `Authorities`. And we can expose `authorities` details to the public view. Only `student` data is kept hidden from the api. Since `students` can only comment and not post any solution we can do this.

drigio commented on services/IssueService.js:
```diff
@@ -0,0 +1,159 @@
+      const issues = await Issue.find({
+        isInappropriate: false,
+        isDeleted: false,
+      }).sort("-upvotes");
```
I guess we should pass the sorting filters like `upvotes` or `date` via a parameter and keep `upvotes` as default maybe?

FaiyazMujawar replied:
I think this should be further discussed.

drigio commented on services/IssueService.js:
```diff
@@ -0,0 +1,159 @@
+        issue.isInappropriate =
+          updatedIssue.isInappropriate || issue.isInappropriate;
+
+        if ((userID = updatedIssue.upvoters)) {
```
Doesn't `updatedIssue.upvoters` return an `array of zprns`?  How are we fetching a single `userId` from it then?

FaiyazMujawar replied:
At a time, in each call, the updated issue would contain a single upvoter; this was my consideration. It was named so as to maintain consistency. Would change if required.

drigio replied:
Okay let's see what you have got on your mind.

FaiyazMujawar replied:
We should discuss this further for clarification.

drigio commented on services/IssueService.js:
```diff
@@ -0,0 +1,159 @@
+          } else {
+            // Else, add the userID to upvoters list.
+            issue.upvotes += 1;
+            issue.upvoters.push(userID);
```
Here as well, I guess you are pushing a complete array here. Because `upvoters` will return an array rather than a single `userId`

FaiyazMujawar replied:
Same as above.

drigio commented on test/unit/services/IssueService.test.js:
```diff
@@ -0,0 +1,211 @@
+
+describe("IssueService", () => {
+  // Todo: Add tests for getAllIssues() & getAllIssuesByDepartment()
+  //? Not getting how to test the chained mongoose methods
```
Didn't we find a way for these `chained` methods? This [SO ans](https://stackoverflow.com/a/52359694) helps. We can use the `callsFake()` method for mocking.

FaiyazMujawar replied:
I tried `callsFake` but it didn't work.

drigio replied:
Let's see this. Ill pull in this PR and run tests

drigio commented on test/unit/services/IssueService.test.js:
```diff
@@ -0,0 +1,211 @@
+      upvoters: [],
+      comments: [],
+      solutions: [],
+      createdBy: "userID",
```
`createdBy` is of type `number`.

FaiyazMujawar replied:
Oh, my bad!

drigio commented on test/unit/services/IssueService.test.js:
```diff
@@ -0,0 +1,211 @@
+          "userID"
+        );
+        expect(create.calledOnce).to.be.true;
+        expect(issue).to.equal(newIssue);
```
This should fail as `createdBy` will not have anything in `issue` as the `typecasting` will fail.

FaiyazMujawar replied:
Will definitely look into it

drigio commented on test/unit/services/IssueService.test.js:
```diff
@@ -0,0 +1,211 @@
+        expect(updatedIssue).to.be.equal(issue);
+        expect(findOne.calledOnce).to.be.true;
+        expect(save.calledOnce).to.be.true;
+      } catch (error) {
```
Yes. `createdBy` is of type `number` hence most probably the error.

drigio commented on test/unit/services/IssueService.test.js:
```diff
@@ -0,0 +1,211 @@
+      } catch (error) {
+        expect(findOne.calledOnce).to.be.true;
+
+        //! [BUG] save.calledOnce should be true, but is false.
```
Will need to check this once after running tests.

drigio reviewed (commented):
Everything else looks in place. Also could you mention `tests` which are remaining. Give exhaustive list.

FaiyazMujawar commented:
> Everything else looks in place. Also could you mention `tests` which are remaining. Give exhaustive list.

The tests for `getAllIssuesByDepartment` and `getAllIssues` are not written because of the chained methods. Rest all the tests cases are written, as far as I can think. Please add some more if required.

drigio commented on models/Comment.js:
```diff
@@ -0,0 +1,9 @@
+class Comment {
+  constructor(userID, solution) {
+    this.postedBy = userID;
+    this.solution = solution;
```
Why does `comment` have a `solution` field?

drigio commented:
I guess all the `tests` are now passing? Will work on the remaining tests if all is good to go.

FaiyazMujawar commented:
@drigio No, they are still failing 🤦

drigio commented:
Update commit names to make them follow our commit guidelines.

FaiyazMujawar commented:
> Update commit names to make them follow our commit guidelines.

Which commit name?

drigio commented:
Read your commit names in this PR.
`Implemented controller`, etc
Also what about test cases? If they keep failing its no use then.

drigio commented on models/roles.js:
```diff
@@ -5,14 +5,20 @@ exports.roles = (() => {
+    .createAny("comment");
 
-  ac.grant("student_moderator").extend("student");
+  ac.grant("student_moderator").extend("student").deleteAny("issue");
```
I think we shouldn't give this right to student moderator. Rather only marking the issue as `inappropriate`

drigio commented on models/roles.js:
```diff
@@ -5,14 +5,20 @@ exports.roles = (() => {
     .updateOwn("profile")
-    .deleteOwn("profile");
+    .deleteOwn("profile")
+    .createOwn("solution");
```
Can't `auth_level_one` comment on issues?

drigio commented on routes/issue.js:
```diff
@@ -0,0 +1,78 @@
+);
+
+// Delete an issue
+router.delete("/:id", allowIfLoggedIn, hasAccessTo(), deleteIssue);
```
`hasAccessTo()` has no parameters here.

drigio reviewed (changes requested):
Left a few comments. Just have a look at them, then we can merge this in.

FaiyazMujawar committed:
issues: implemented CRUD on issues

In this:
- Implemented database schema for issues.
- Implemented service layer for issues.
- Implemented controller for issues.

drigio commented:
Looks good for now. Need to create tests for CI. Create an issue for adding tests so that we can track it side by side. Let me know if there are any further comments on it, so that I can merge it in the codebase.

FaiyazMujawar commented:
> Looks good for now. Need to create tests for CI. Create an issue for adding tests so that we can track it side by side. Let me know if there are any further comments on it, so that I can merge it in the codebase.

Nope. Please go ahead and merge.
